### PR TITLE
trace: Migrate to sentry-go

### PR DESCRIPTION
**Pull request by keegancsmith**
raven-go has been deprecated for a while. This migrates us to the blessed
package for sending events to Sentry. It works in a strange way for Go were it
seems to simulate goroutine local storage. Notice that the `ConfigureScope`
and `CaptureException` functions both do not rely on any sort of
`context`. There doesn't seem to be any alternatives which work with less
magic :(

Additionally sentry-go always does some work, even if a request doesn't report
an exception. It will always clone a `hub` for a request, and will always
create a `sentry.Request` (with a copy of headers).

This commit also I removes the `reportError` function from handlerutil. It
didn't seem to get used in practice (our production sentry has no hits for the
tags it sets).

Additionally we move some "tags" to "extras". Tags are better suited for low
cardinality values (the values have stats calculated on them).

Finally I improve the initialization such that we only configure sentry if
sentry related configuration changes.

Part of https://github.com/sourcegraph/sourcegraph/issues/6738

### CODE REVIEW

**Comment by keegancsmith** [comment]
@sourcegraph/core-services please read the description, there are some concerns I have with this library which makes me consider sticking to the deprecated client. However, whatever extra work we do for this instrumentation is likely tiny compared to the rest of the request / opentracing.

I got started on this because I wanted to improve error reporting in sentry (still sucks) and noticed we had a very outdated library. Using this will only slightly improve our sentry reporting, there will be follow-up work to try and use better stack traces in sentry (`pkg/errors` will hopefully give us better errors than sentry is currently reporting)

**Comment by tsenart** [comment]
Here's an alternative proposal: Since we only use this for our own environments, how about we integrate with [Stackdriver Errors](https://cloud.google.com/error-reporting/docs/setup/go) instead?

**Comment by slimsag** [comment]
I am highly in favor of anything that customers could also use without purchasing some additional product license. Is there anything even more general purpose we could use? (Logs, something not tied to GCP, etc?)

**Comment by tsenart** [comment]
> (Logs, something not tied to GCP, etc?)

Good point! I think if we shipped something like https://grafana.com/oss/loki/ with Sourcegraph, we could fare without a dedicated error reporting tool.

**Comment by keegancsmith** [comment]
We tried out loki a while ago, it worked pretty poorly if I am not mistaken. But it was still early days for the loki project (cc @ggilmore).

Sentry is great, and I believe we have some customers that have integrated it before. Additionally we use it to record exceptions in our typescript stack, and I believe the web team relies on that a lot.

The issue here is that `sentry-go` seems to have a few issues / strange model likely copied from another language rather than working how a go programmer would expect. However, at the end of the day the event we send to the sentry server is a good idea and simple + the sentry UI is world class. So I am in favour with sticking with sentry unless we have a good reason against it. I can wield the sentry client to be acceptable for us, and if I have any stamina left after that I may give them some constructive feedback.

File: internal/trace/httptrace.go [hunk 1]
```diff
@@ -165,20 +179,27 @@ func Middleware(next http.Handler) http.Handler {
...
+					"duration": m.Duration.String(),
+				})
+				scope.SetTags(map[string]string{
+					"code":          strconv.Itoa(m.Code),
+					"routename":     routeName,
+					"userAgent":     r.UserAgent(),
+					"xForwardedFor": r.Header.Get("X-Forwarded-For"),
+					"graphql_error": strconv.FormatBool(gqlErr),
+				})
+				sentry.CaptureException(requestErrorCause)
```
Comment (unknwon) [inline internal/trace/httptrace.go]: Hmm, I think this block of code isn't doing what we want. Call to the `sentry.ConfigureScope` seems like using global state thus isn't made for request-level (as you mentioned there is no concept of `context`). 

https://github.com/getsentry/sentry-go/blob/759fbf8e9221cd2a2c3a26e3d7b41ecae2ed89d9/hub.go#L78-L81

```go
// CurrentHub returns an instance of previously initialized `Hub` stored in the global namespace.
func CurrentHub() *Hub {
	return currentHub
}
```

Later I found out the "hub" object we actually want to use is in the request context (as we call `HandleFunc` in line 202):

https://github.com/getsentry/sentry-go/blob/759fbf8e9221cd2a2c3a26e3d7b41ecae2ed89d9/http/sentryhttp.go#L70-L75

```go
		ctx := sentry.SetHubOnContext(
			r.Context(),
			hub,
		)
		defer h.recoverWithSentry(hub, r)
		handler(rw, r.WithContext(ctx))
```

Therefore, inside this `h` function, we should get our `hub` via [`sentry. GetHubFromContext`](https://github.com/getsentry/sentry-go/blob/HEAD/hub.go#L344).

**Comment by slimsag** [comment]
> Good point! I think if we shipped something like https://grafana.com/oss/loki/ with Sourcegraph, we could fare without a dedicated error reporting tool.

Agreed, I am strongly in favor of investigating this.

> We tried out loki a while ago, it worked pretty poorly if I am not mistaken. But it was still early days for the loki project (cc @ggilmore).

It has improved a lot since then. See also this HN comment chain: https://news.ycombinator.com/item?id=21953930

> Sentry is great, and I believe we have some customers that have integrated it before.

Sentry is great -- but we have not had any customers that have integrated it for Sourcegraph usage. We've had customers that used Sentry for _their_ applications, and wanted error generation sites in Sourcegraph's code view to link to their Sentry -- but we've never convinced a customer to set up Sentry with Sourcegraph for us to debug our application. And given it is a commercial offering, I think that would be very difficult.

> Additionally we use it to record exceptions in our typescript stack, and I believe the web team relies on that a lot.

This is true, but means we have zero insight into customer deployments, and is the main reason I suggest we stop using Sentry. When a customer runs into an issue and suddenly all of your debugging tooling is gone.. well, that issue isn't going to be easy to fix.

**Comment by keegancsmith** [comment]
> > We tried out loki a while ago, it worked pretty poorly if I am not mistaken. But it was still early days for the loki project (cc @ggilmore).
> 
> It has improved a lot since then. See also this HN comment chain: https://news.ycombinator.com/item?id=21953930

Great to hear. Looking forward to some unification here.
 
> > Sentry is great, and I believe we have some customers that have integrated it before.
> 
> Sentry is great -- but we have not had any customers that have integrated it for Sourcegraph usage. We've had customers that used Sentry for _their_ applications, and wanted error generation sites in Sourcegraph's code view to link to their Sentry -- but we've never convinced a customer to set up Sentry with Sourcegraph for us to debug our application.

I may be remembering incorrectly, but a very long time ago we made the sentry details configurable for https://app.hubspot.com/contacts/2762526/company/464956351/ (cc @beyang).

> And given it is a commercial offering, I think that would be very difficult.

It is OSS, just with a hosted option you pay for. https://github.com/getsentry/sentry (Disclosure: I've setup sentry internally for companies a few times and also contributed to it. Was many years ago though.)

> > Additionally we use it to record exceptions in our typescript stack, and I believe the web team relies on that a lot.
> 
> This is true, but means we have zero insight into customer deployments, and is the main reason I suggest we stop using Sentry. When a customer runs into an issue and suddenly all of your debugging tooling is gone.. well, that issue isn't going to be easy to fix.

I think the @sourcegraph/web  team relies on sentry. So before we rip out the frontend bit we should consult with them. Note: Sentry's API is just sending a JSON payload, which is compatible with how loki works. Sentry just does nice stuff around de-duplication/reporting.

But given how poor our backend integration is I'd be happy removing support for it in the backend. And for now just closing this issue out.

**Pull request closed without merging**